### ISSUE TITLE
[Modernization] Remove all the 1.7- code!

### ISSUE THREAD

**Issue by srikanth-sankaran** [eclipse-jdt/eclipse.jdt.core#3304]
I started this ticket with the title: `Remove CLDC support`  with the description:

`org.eclipse.jdt.internal.compiler.impl.CompilerOptions.UNSUPPORTED_VERSIONS` lists `VERSION_CLDC1_1` as unsupported.
We still have various vestiges of this in ECJ: for instance STACK_MAP attribute ...

But larger question though is what do we want to do with all the < 1.8 code in the compiler ? I think like we did for $IDENTITY-COMPARISON$ annotation, we should perhaps come up with an instrumented compiler that will flag all unreachable blocks and then use quick fixes to prune them after eyeballing the changes.

I have seen some observations against aggressively purging old code - but would like a discussion that - Done early in a release in the first week of M1, it can be managed well methinks.

I personally would like see all the clutter gone!

**Comment by srikanth-sankaran** [comment]
@jarthana @mpalat @stephan-herrmann @jukzi @iloveeclipse - Any reason why we MUST not do this - as outlined in a very early stage of M1 of an upcoming release ? Mechanically flagged unreachable blocks, but manually deleted via quick fixes after inspection and reviewed before commit ?

**Comment by srikanth-sankaran** [comment]
https://bugs.eclipse.org/bugs/show_bug.cgi?id=417803 

> I think like we did for the IDENTITY-COMPARISON 
>  annotation, we should perhaps come up with an instrumented compiler that will flag all unreachable blocks and then use quick fixes to prune them after eyeballing the changes.

https://bugs.eclipse.org/bugs/show_bug.cgi?id=417803 is the relevant ticket

**Comment by srikanth-sankaran** [comment]
(It may be tricky to catch all the dead blocks through this mechanical approach cf. null analysis & correlation related issue)

**Comment by stephan-herrmann** [comment]
Let me remind you, that we are not yet done with the previous step, i.e., disabling compliance levels below 8. Let's first resolve #2758 and #3297 before we start discussing the next step.

That said, I see no problem with removing a few individual code blocks in areas that we actively work on, preferably with a code review involved.

**Comment by stephan-herrmann** [comment]
Just so you have my opinion on the table, too:
* I'm much in favor of small step clean-up, well interleaved with feature work, rather than stopping the world until all is 100% neat.
* I am not in favor of calling a clean-up done before it is 100% complete, including 100% of tests
* I'm not in favor of starting another round of clean-up before the previous round has completed.
* I don't plan to spend much time in reviewing big-scale clean-ups.
* The case of `$Identity-Comparison$` had high urgency, because we made a fundamental design change, where inappropriate `==` was a bug, not just an inconvenience.
* (And large changes always drain some of my time away from JDT to keep my fork alive)

This said so you don't forget I'm the old conservative PITA that I am ... but I'm neither the majority nor do I have the money to support my strategy.

Positively speaking, I, too, will be happy to kick out unused stuff, when and where it gets in my way.

**Comment by stephan-herrmann** [comment]
> The argumentation that we should only start new cleanup after previous one is 100% done is not correct.

"correct" is a funny word in this sentence, but let me rephrase my position: a clean-up should leave us in a better state than before, which implies it should not cause any code decay of its own.

**Comment by srikanth-sankaran** [comment]
> So if we want thishttps://github.com/[/issues/2758](https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2758) to be fixed, please help! But if no one will do that - should we stop now any refactorings? Surely not!

I'll take over this one, although I will work on this in a very gradual fashion

**Comment by stephan-herrmann** [comment]
> It would not be my preference to combine this cleanup in a commit for feature work or bug fixes and pollute the delta's - not saying you are suggesting that.

If the clean-up part of any given feature change is significant in size, it would surely be a good idea to keep it as a separate commit (and then refrain from squashing on merge).

**Comment by srikanth-sankaran** [comment]
> > It would not be my preference to combine this cleanup in a commit for feature work or bug fixes and pollute the delta's - not saying you are suggesting that.
> 
> If the clean-up part of any given feature change is significant in size, it would surely be a good idea to keep it as a separate commit (and then refrain from squashing on merge).

That would help but not entirely. Looking at the delta between two versions spaced apart could bring up these "noise" changes that went in their own non-squashed commits. They will anyway but more likely when it is done in installments,

**Comment by stephan-herrmann** [comment]
> So if we want thishttps://github.com/[/issues/2758](https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2758) to be fixed, please help!

Did you see that one of those tests was already ticked off? Guess by whom :)
And during the course we even got https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2727 resolved, haven't you been waiting for that?
